### ISSUE TITLE
linemod kinect magic numbe

### ISSUE THREAD

**Issue by bhack** [Itseez/opencv_contrib#208]
I still see in linemod.cpp
>@todo Magic number 1150 is focal length? This is something like
        /// f in SXGA mode, but in VGA is more like 530.

@vrabaud Do you know why we still have this fixed parameter in linemod? Is it real the focal length?

**Comment by vrabaud** [comment]
yes it is. This was a hack actually: an approximation of the normal. I actually re-did the computation following the paper (Stefan told me that was actually the way to go to get higher accuracy): https://github.com/Itseez/opencv_contrib/blob/master/modules/rgbd/src/normal.cpp#L472

**Comment by bhack** [comment]
Thank you for point there.  I missed this code in contrib.  I see that there you  already use the camera matrix. So that can be ready to be used with other depth sensors.

**Comment by mtamburrano** [comment]
Hi Vincent,
maybe I'm wrong, but I'm not sure about the computation of the normals is correct for Linemod...

first of all, seems like [here][1] suggests to use CV_16U depth images for Linemod, and it is fine, but [this assert][2] and [this][3] don't allow it.

Anyway, even fixing the asserts, the resulting Mat seems to me broken, I get an unsigned short matrix (the normals Mat is supposed to be the same type as the depth? Why?) mostly filled with 0 and capped unsigned short values.

This is the code I'm using to test it after have fixed the asserts, passing CV_32F instead of 16U in the constructor leads to a normals mat with mostly NaN values.
Let me know if I'm doing something wrong:

```
Matx33f K_ = Matx33f::eye(); //camera matrix
K_(0,0) = 525.0;
K_(1,1) = 525.0;
K_(0,2) = 319.5;
K_(1,2) = 239.5;

rgbd::RgbdNormals linemod_normals(480, 640, CV_16U, K_, 3, rgbd::RgbdNormals::RGBD_NORMALS_METHOD_LINEMOD);
Mat normals;
linemod_normals.operator ()(depth, normals); //depth is a CV_16U Mat

```
I don't transform the depth into 3d points because the METHOD_LINEMOD seems to accept also the raw depth. I've also tried to pass the 3dPoints instead with no significative differences

[1]: https://github.com/Itseez/opencv_contrib/blob/master/modules/rgbd/include/opencv2/rgbd.hpp#L130
[2]: https://github.com/Itseez/opencv_contrib/blob/master/modules/rgbd/src/normal.cpp#L683
[3]: https://github.com/Itseez/opencv_contrib/blob/master/modules/rgbd/src/normal.cpp#L733

**Comment by bhack** [comment]
@vrabaud I've not checked normals compute in detail but at a first look seems to me that there is something that could not work between templates, 16U handling, and ASSERTS.

**Comment by vrabaud** [comment]
it seems the linemod normal code can handle different depth types:
https://github.com/Itseez/opencv_contrib/blob/master/modules/rgbd/src/normal.cpp#L375

Now, just add the 16U to the tests for linemod (https://github.com/Itseez/opencv_contrib/blob/master/modules/rgbd/test/test_normal.cpp#L262) and you should see errors popping up.

**Comment by vrabaud** [comment]
All that should be fixed (I believe), should be: the test (to also handle 16U for LINE-MOD), the Assert, the doc of the function.

**Comment by bhack** [comment]
@vrabaud Also in the actual unmodified version  opencv_test_rgbd exit with a segmentation fault. See https://github.com/Itseez/opencv_contrib/blob/master/modules/rgbd/test/test_normal.cpp#L122:L130

**Comment by vrabaud** [comment]
ok, after looking at it: code is good but documentation is wrong (it does not allow CV_16U as output normals: it does not make sense anyway). Tests do not crash for me: ```./bin/opencv_test_rgbd --gtest_filter=*Normals*```

Odometry tests crash for me though.

**Comment by bhack** [comment]
Ok tomorrow I will check the filtered test on Normals cause I only runned the whole module test. Probably it is caused by Odometry.

**Comment by bhack** [comment]
@vrabaud I've checked also the filtered test. But still segfault at rgbd/test/test_normal.cpp:130

**Comment by bhack** [comment]
Ok,
we have workaround templates for now to check result and works fine. But depth normal parameters seems to assert 32F or 64F for normals but why [here](https://github.com/Itseez/opencv_contrib/blob/master/modules/rgbd/src/normal.cpp#L378) in Linemod case seems to use long (16S) for linemod normals?

**Comment by vrabaud** [comment]
no, the LINEMOD class is templated on the normal type (32 or 64: that's all it should support ).
The template arguments you mention are the depth of the depth image (and LINEMOD also works with 16S cf https://github.com/Itseez/opencv_contrib/blob/master/modules/rgbd/src/normal.cpp#L807), and a container for computation (for internal purpose).

Nothing needs to be workaround, just the doc fixed (and a test for a 16U depth image for LINEMOD).

**Comment by mtamburrano** [comment]
Hi,
just sent a PR here #217.
I hope the tests are fine

**Comment by michael-peng** [comment]
Hi,
If you change the magic number 1150 to 530, it will change the readout from the look up table, will it affect the matching result? By the way, can anyone tell me how to generate the look up table?
Many thanks.
